internal/handler/msg_updateuser.go

```diff
@@ -29,6 +38,201 @@ func (h *Handler) MsgUpdateUser(ctx context.Context, msg *wire.OpMsg) (*wire.OpM
+		var ce *handlererrors.CommandError
+		if errors.As(err, &ce) && ce.Code() == handlererrors.ErrBadValue {
+			return nil, handlererrors.NewCommandErrorMsg(
+				handlererrors.ErrMissingField,
+				"BSON field 'updateUser.roles' is missing but a required field",
+			)
+		}
+
+		return nil, lazyerrors.Error(err)
+	}
+
+	if err = common.UnimplementedNonDefault(document, "roles", func(v any) bool {
+		r, ok := v.(*types.Array)
+		return ok && r.Len() == 0
+	}); err != nil {
+		return nil, err
+	}
+
+	common.Ignored(document, h.L, "writeConcern", "authenticationRestrictions", "comment")
+
+	defMechanisms := must.NotFail(types.NewArray())
+
+	mechanisms, err := common.GetOptionalParam(document, "mechanisms", defMechanisms)
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	if mechanisms != nil {
+		iter := mechanisms.Iterator()
+		defer iter.Close()
+
+		for {
+			var v any
+			_, v, err = iter.Next()
+
+			if errors.Is(err, iterator.ErrIteratorDone) {
+				break
+			}
+
+			if err != nil {
+				return nil, lazyerrors.Error(err)
+			}
+
+			if v != "PLAIN" {
+				return nil, handlererrors.NewCommandErrorMsg(
+					handlererrors.ErrBadValue,
+					fmt.Sprintf("Unknown auth mechanism '%s'", v),
+				)
+			}
+		}
+	}
+
+	var credentials *types.Document
+	if document.Has("pwd") {
+		credentials = types.MakeDocument(0)
+		pwdi := must.NotFail(document.Get("pwd"))
+		pwd, ok := pwdi.(string)
+
+		if !ok {
+			return nil, handlererrors.NewCommandErrorMsg(
+				handlererrors.ErrTypeMismatch,
+				fmt.Sprintf("BSON field 'updateUser.pwd' is the wrong type '%s', expected type 'string'",
+					handlerparams.AliasFromType(pwdi)),
+			)
+		}
+
+		if pwd == "" {
+			return nil, handlererrors.NewCommandErrorMsg(
+				handlererrors.ErrSetEmptyPassword,
+				"Password cannot be empty",
+			)
+		}
+
+		credentials.Set("PLAIN", must.NotFail(password.PlainHash(pwd)))
+	}
+
+	usersCol, err := adminDB.Collection("system.users")
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	var filter *types.Document
+	filter, err = usersInfoFilter(false, false, "", []usersInfoPair{
+		{
+			username: username,
+			db:       dbName,
+		},
+	})
+
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	// Filter isn't being passed to the query as we are filtering after retrieving all data
+	// from the database due to limitations of the internal/backends filters.
+	qr, err := usersCol.Query(ctx, nil)
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	defer qr.Iter.Close()
+
+	var saved *types.Document
+
+	for {
+		var v *types.Document
+		_, v, err = qr.Iter.Next()
+
+		if errors.Is(err, iterator.ErrIteratorDone) {
+			break
+		}
+
+		if err != nil {
+			return nil, lazyerrors.Error(err)
+		}
+
+		var matches bool
+		matches, err = common.FilterDocument(v, filter)
+
+		if err != nil {
+			return nil, lazyerrors.Error(err)
+		}
+
+		if matches {
+			saved = v
```

Comment:
do also we need to close the iterator or exit from the loop or something?

Reply:
Iterator is deferred closed on preceding line `qr.Iter.Close()`.

This loop should always contain a maximum 1 item, but I'm adding a `break` just to make it more explicit.

Reply:
Hmm, could you help me understand why this loop should always contain a maximum 1 item? Doesn't it contain all the users?

Reply:
Oops! You're correct. Please disregard my comment.

It should contain at maximum one item, but we're not filtering at the database layer at this moment.